### PR TITLE
Add memory flush operations for dash::NArray

### LINKED ISSUE #312: Add support for memory flush in NArray
Currently there is no possibility to explicitly issue a flush on the globmem of a `dash::NArray / Matrix`.

### PR DESCRIPTION
**Pull request by fuchsto**
Fixes #312

### CODE REVIEW

File: dash/include/dash/Array.h [hunk 1]
```diff
@@ -1172,6 +1172,38 @@ class Array
...
+   */
+  inline void flush() const {
+    m_globmem->flush();
+  }
+
+  /**
+   * Complete all outstanding non-blocking operations executed by the
+   * local unit on the array's underlying global memory.
+   */
+  inline void flush_local() const {
```
Comment (fmoessbauer) [inline dash/include/dash/Array.h]: Please document which flushes are collective.

File: dash/include/dash/matrix/internal/Matrix-inl.h [hunk 2]
```diff
@@ -306,6 +306,34 @@ ::barrier() const {
 }
 
 template <typename T, dim_t NumDim, typename IndexT, class PatternT>
+inline void
+Matrix<T, NumDim, IndexT, PatternT>
+::flush() {
```
Comment (fmoessbauer) [inline dash/include/dash/matrix/internal/Matrix-inl.h]: Same here

File: dash/include/dash/memory/GlobStaticMem.h [hunk 3]
```diff
@@ -402,28 +402,34 @@ class GlobStaticMem
   }
 
   /**
-   * Complete all outstanding asynchronous operations on the referenced
-   * global memory on all units.
+   * Complete all outstanding non-blocking operations executed by all units.
```
Comment (fmoessbauer) [inline dash/include/dash/memory/GlobStaticMem.h]: And here

File: dash/include/dash/Array.h (hunk 1 above)
Comment (fuchsto) [reply dash/include/dash/Array.h]: Flushes are not collective.

**Comment by fuchsto** [comment]
@fmoessbauer I don't know what you mean by "collective flushes". Blocking, yes, but collective?

**Comment by rkowalewski** [comment]
hmm what to review there are no file changes?

**Comment by fuchsto** [comment]
Already merged as part of https://github.com/dash-project/dash/pull/341

**Pull request closed without merging**